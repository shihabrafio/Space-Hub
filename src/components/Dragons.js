import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  fetchDragons,
  reserveDragon,
  cancelReservation,
} from '../redux/dragons/dragonsSlice';
import '../App.css';

const Dragons = () => {
  const { dragons, status, error } = useSelector((state) => state.dragons);
  const dispatch = useDispatch();

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchDragons());
    }
  }, [status, dispatch]);

  if (status === 'loading') {
    return <h2 style={{ textAlign: 'center' }}>Loading...</h2>;
  }

  if (status === 'failed') {
    return <h3 style={{ textAlign: 'center' }}>{error}</h3>;
  }

  return (
    <section className="container">
      {dragons.map((dragon) => (
        <div className="card" key={dragon.id}>
          <div className="img">
            <img src={dragon.flickr_images} alt="123" />
          </div>
          <div className="card-content">
            <h3>{dragon.name}</h3>
            <p className="desc">
              {dragon.reserved && <span className="badge">Reserved</span>}
              {dragon.description}
            </p>
            {!dragon.reserved ? (
              <button
                className="btn primary-btn"
                type="button"
                onClick={() => dispatch(reserveDragon(dragon.id))}
              >
                Reserve Dragon
              </button>
            ) : (
              <button
                className="btn secondary-outline-btn"
                onClick={() => dispatch(cancelReservation(dragon.id))}
                type="button"
              >
                Cancel Reservation
              </button>
            )}
          </div>
        </div>
      ))}
    </section>
  );
};

export default Dragons;
