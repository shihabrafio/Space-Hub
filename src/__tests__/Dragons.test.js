import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
import Dragons from '../components/Dragons';
import dragonsReducer from '../redux/dragons/dragonsSlice';

describe('Dragons component', () => {
  it('renders correctly', () => {
    const dragons = [
      {
        id: 1,
        name: 'Dragon 1',
        flickr_images: 'dragon1.jpg',
        description: 'Dragon 1 Description',
        reserved: false,
      },
      {
        id: 2,
        name: 'Dragon 2',
        flickr_images: 'dragon2.jpg',
        description: 'Dragon 2 Description',
        reserved: true,
      },
    ];

    const rootReducer = combineReducers({
      dragons: dragonsReducer,
    });

    const store = configureStore({
      reducer: rootReducer,
      preloadedState: {
        dragons: {
          dragons,
          status: 'success',
          error: null,
        },
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Dragons />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
