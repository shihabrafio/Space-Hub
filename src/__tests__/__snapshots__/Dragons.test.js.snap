// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons component renders correctly 1`] = `
<div>
  <section
    class="container"
  >
    <div
      class="card"
    >
      <div
        class="img"
      >
        <img
          alt="123"
          src="dragon1.jpg"
        />
      </div>
      <div
        class="card-content"
      >
        <h3>
          Dragon 1
        </h3>
        <p
          class="desc"
        >
          Dragon 1 Description
        </p>
        <button
          class="btn primary-btn"
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="img"
      >
        <img
          alt="123"
          src="dragon2.jpg"
        />
      </div>
      <div
        class="card-content"
      >
        <h3>
          Dragon 2
        </h3>
        <p
          class="desc"
        >
          <span
            class="badge"
          >
            Reserved
          </span>
          Dragon 2 Description
        </p>
        <button
          class="btn secondary-outline-btn"
          type="button"
        >
          Cancel Reservation
        </button>
      </div>
    </div>
  </section>
</div>
`;
